crates_io_tarball/src/manifest.rs

```diff
@@ -1,44 +1,89 @@
-use cargo_toml::OptionalFile;
-use derive_deref::Deref;
-use serde::{de, Deserialize, Deserializer};
-
-#[derive(Debug, Deserialize)]
-pub struct Manifest {
-    #[serde(alias = "project")]
-    pub package: Package,
-}
-
-#[derive(Debug, Deserialize)]
-#[serde(rename_all = "kebab-case")]
-pub struct Package {
-    pub name: String,
-    pub version: String,
-    #[serde(default)]
-    pub readme: OptionalFile,
-    pub repository: Option<String>,
-    pub rust_version: Option<RustVersion>,
-}
-
-#[derive(Debug, Deref)]
-pub struct RustVersion(String);
-
-impl PartialEq<&str> for RustVersion {
-    fn eq(&self, other: &&str) -> bool {
-        self.0.eq(other)
-    }
-}
-
-impl<'de> Deserialize<'de> for RustVersion {
-    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<RustVersion, D::Error> {
-        let s = String::deserialize(d)?;
-        match semver::VersionReq::parse(&s) {
-            // Exclude semver operators like `^` and pre-release identifiers
-            Ok(_) if s.chars().all(|c| c.is_ascii_digit() || c == '.') => Ok(RustVersion(s)),
-            Ok(_) | Err(..) => {
-                let value = de::Unexpected::Str(&s);
-                let expected = "a valid rust_version";
-                Err(de::Error::invalid_value(value, &expected))
-            }
-        }
+use cargo_toml::{Dependency, DepsSet, Error, Inheritable, Manifest, Package};
```

Comment:
Is the extra validation worth it?

I'm concerned about us (very easily) adding new things that will then fail on publish.  The person publishing would then need to figure out it is a bug in crates.io, report it, someone respond, create a PR upstream with `cargo_toml`, wait for a fix and publish, then updates and deploy crates.io.

Reply:
Would it be reasonable to add logging / alerting such that we could detect an increase in failed toml parsing?

Reply:
yep, that seems like a reasonable idea :)

Reply:
> I'm concerned about us (very easily) adding new things that will then fail on publish.

Wouldn't a breaking change like that cause bigger problems anyway, since crates.io is by no means the only downstream parser and consumer of cargo manifests?

Reply:
I'm not even talking about breaking backwards compatibility but breaking forward compatibility.

Think of cases like:
- Add new supported values
- Add new supported types

Cargo needs to have a strict parser for these because an unknown value could have a major impact on behavior.  If another tool is acting off of the field, they should hard error as well.    However, we need to be able to add new values or else cargo is stuck as-is.  This is possible to support in serde using a catch-all variant.  `cargo_toml` instead [errors](https://docs.rs/cargo_toml/0.15.3/src/cargo_toml/cargo_toml.rs.html#651) on unknown values.

A trickier one is adding new types which we've done but [`cargo_metadata`](https://github.com/oli-obk/cargo_metadata/issues/240) also has hard errors for that.

So assuming we only maintain backward compatibility.  Most likely we'll be playing whack-a-mole with `cargo_toml` and [`cargo_metadata`](https://github.com/oli-obk/cargo_metadata/issues/240) for them to properly handle new values being emitted.

We then have to deal with making sure they are updated when the new values are out (at minimum, a strong communication path).

Tying this back to this PR, making most of `cargo_toml` loosey-goosey for compatibility removes most of the validation that crates-io can do.

Independent of all of that., in my opinion, is that pushing responsibility for this stuff to third-parties has made it so we are unaware of the problems their users face and it is easier to make [innocuous changes that break them](https://github.com/rust-lang/cargo/issues/10543)